go-controller/pkg/ovn/controller/apbroute/external_controller_policy.go

```diff
@@ -3,377 +3,240 @@ package apbroute
-}
 
-func (m *externalPolicyManager) hasPolicyInNamespace(policyName, namespaceName string) bool {
-
-	nsInfo, found := m.getNamespaceInfoFromCache(namespaceName)
-	if !found {
-		klog.V(5).Infof("Namespace %s not found while consolidating namespaces using policy %s", namespaceName, policyName)
-		return false
-	}
-	defer m.unlockNamespaceInfoCache(namespaceName)
-	return nsInfo.Policies.Has(policyName)
-}
-
-// locks the namespace cache and executes discrepancyFunc
-func (m *externalPolicyManager) processPolicyDiscrepancyInNamespace(nsName string, routePolicy *adminpolicybasedrouteapi.AdminPolicyBasedExternalRoute,
-	discrepancyFunc func(string, *adminpolicybasedrouteapi.AdminPolicyBasedExternalRoute, *namespaceInfo) error) error {
-	cacheInfo, found := m.getNamespaceInfoFromCache(nsName)
-	if !found {
-		klog.V(5).Infof("Namespace %s not found in cache, creating", nsName)
-		cacheInfo = m.newNamespaceInfoInCache(nsName)
-	}
-	defer m.unlockNamespaceInfoCache(nsName)
-	return discrepancyFunc(nsName, routePolicy, cacheInfo)
-}
-
-func (m *externalPolicyManager) reconcilePolicyWithNamespacesAndPods(routePolicy *adminpolicybasedrouteapi.AdminPolicyBasedExternalRoute) error {
-
-	nss, err := m.listNamespacesBySelector(&routePolicy.Spec.From.NamespaceSelector)
-	if err != nil {
-		return err
-	}
-	targetNs := sets.New[string]()
-	for _, ns := range nss {
-		targetNs.Insert(ns.Name)
-	}
-
-	klog.V(4).Infof("Reconciling namespaces %+v for policy %s", strings.Join(targetNs.UnsortedList(), ", "), routePolicy.Name)
-	// reconcile Namespaces
-	currentNs, err := m.listNamespacesWithPolicy(routePolicy.Name)
-	if err != nil {
-		return err
-	}
-	klog.V(5).Infof("List of existing namespaces with policy %s: %s", routePolicy.Name, strings.Join(currentNs.UnsortedList(), ", "))
-	if len(currentNs.Difference(targetNs)) > 0 {
-		klog.V(4).Infof("Removing APB policy %s in namespaces %+v", routePolicy.Name, currentNs.Difference(targetNs))
-	}
-	// namespaces where the policy should not be applied anymore.
-	for nsName := range currentNs.Difference(targetNs) {
-		err = m.processPolicyDiscrepancyInNamespace(nsName, routePolicy, m.removePolicyFromNamespace)
-		if err != nil {
-			return err
+		// get existing policy and update routePolicySyncCache
+		existingPolicy, found := m.routePolicySyncCache.Load(policyName)
+		if !found {
+			if updatedPolicy == nil {
+				// nothing to do
+				return nil
+			}
+			existingPolicy = newRoutePolicyState()
+			m.routePolicySyncCache.Store(policyName, existingPolicy)
 		}
-	}
-	if len(targetNs.Difference(currentNs)) > 0 {
-		klog.V(4).Infof("Adding policy %s to namespaces %+v", routePolicy.Name, targetNs.Difference(currentNs))
-	}
-	// namespaces where the policy should now be applied
-	for nsName := range targetNs.Difference(currentNs) {
-		err = m.processPolicyDiscrepancyInNamespace(nsName, routePolicy, m.applyPolicyToNamespace)
-		if err != nil {
-			return err
+
+		err = m.updateRoutePolicy(existingPolicy, updatedPolicy)
```

Comment:
this path is complicated to follow, maybe some more comments would help clarify how all these different building blocks fit together would be nice. From what I see its:

1. routePolicySyncCache -> holds routePolicyState -> map[string]map[ktypes.NamespacedName]*podInfo -> podInfo holds gateways -> gateway info that also contains if the config was applied
2. syncRoutePolicy takes a lock on the routePolicySyncCache policy name key, and first policyReferencedObjects cache is locked and updated with the kapi resources this policy targets, then unlocked
3. syncRoutePolicy then calls updateRoutePolicy function, and the updatedPolicy is first used to determine what to delete. After those deletions the existingPolicy is updated to reflect the changes.
4. Next the new changes are applied, if not found in existing policy, applyPodConfig is called, and updates the existing policy with the new changes, including status if the gateway was correctly configured for the pod
5. finally the updateRoutePolicy returns, then the updatedPolicy is checked to insert gw IPs and return them to be updated in the status

Is my understanding correct? If so, why at the end do we return the updatedPolicy gwIPs to be used in the status update. Shouldn't we use the existingPolicy since it is now updated, and only include the ones that were succesfully applied or maybe this is a discussion to defer to another PR for status update stuff...

Reply:
Yeah that is all correct, I added some comments, lmk if it is better and if I can improve it some more.

About the returned ips, an error should actually be returned before `gwIPs` are updated (I will fix that), because when error happened, we don't need to return any gwIPs, since status will include error and not a list of applied ips.
But when update was successful, we know that existing policy has the same ips as updatedPolicy, and it is just easier to get them from updatedPolicy. Because updatedPolicy stores gateways in its fields directly, and existingPolicy stores the same config for every every pod.

Reply:
Looks good. I'm going to just leave this comment open cause I'm sure i will come back here to read it again when I forget it next week :)